### PR TITLE
linkcheck: Ignore URLs that respond with non-Unicode content

### LINKED ISSUE #11041: [linkcheck] PDF anchor (`...pdf#anchor`) leads to `'utf-8' codec can't decode byte ...`
### Describe the bug

Related to https://github.com/sphinx-doc/sphinx/issues/7694.

Note that [the query symbol `?` is not required]( https://en.wikipedia.org/wiki/URL#Syntax) when using an anchor (i.e., `#fragment`).

### How to Reproduce

`index.rst`:
```rst
`link1 <https://wci.llnl.gov/sites/wci/files/2020-08/LLNL-SM-654357.pdf?#page=226>`_
`link2 <https://wci.llnl.gov/sites/wci/files/2020-08/LLNL-SM-654357.pdf#page=226>`_
`link3 <https://docs.python.org/3/whatsnew/3.11.html?#whatsnew311-pep654>`_
`link4 <https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-pep654>`_
```

```bash
$ sphinx-build -b linkcheck . build
```

Output:

```
(           index: line    1) ok        https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-pep654
(           index: line    1) redirect  https://docs.python.org/3/whatsnew/3.11.html?#whatsnew311-pep654 - with unknown code to https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-pep654
(           index: line    1) broken    https://wci.llnl.gov/sites/wci/files/2020-08/LLNL-SM-654357.pdf#page=226 - 'utf-8' codec can't decode byte 0xe2 in position 10: invalid continuation byte
(           index: line    1) broken    https://wci.llnl.gov/sites/wci/files/2020-08/LLNL-SM-654357.pdf?#page=226 - 'utf-8' codec can't decode byte 0xe2 in position 10: invalid continuation byte
```

### Environment Information

```text
Platform:              linux; (Linux-6.0.12-arch1-1-x86_64-with-glibc2.36)
Python version:        3.10.8 (main, Nov  1 2022, 14:18:21) [GCC 12.2.0])
Python implementation: CPython
Sphinx version:        5.3.0
Docutils version:      0.19
Jinja2 version:        3.1.2
```

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix

### Purpose
- Improve robustness during anchor-checks, when the `linkcheck` decodes response content.
- ~~Begin to add more validation of parsed HTML~~.

### Detail
- Do not attempt to check for anchors when UTF8 decoding fails; report affected hyperlinks as `ignored`.
- ~~Report hyperlinks as broken when their content fails validation checks, but allow this to be overridden using URI-match patterns.~~

### Relates
- Resolves #11041.

### CODE REVIEW

**Comment by jayaddison** [comment]
@AA-Turner is it OK if we pause on this one for a while?  I think/hope it's a good idea, but I'm slightly concerned that the distinction between lenient and strict parsing might be a bit arbitrary.  I should perhaps move this back into a draft status.

**Comment by jayaddison** [comment]
Another way to phrase that: I'd like to canvas more (and perhaps leaning skeptical) opinions about this before proceeding.

**Comment by jayaddison** [comment]
Thanks!

File: sphinx/builders/linkcheck.py [hunk 1]
```diff
@@ -617,7 +625,7 @@ def _get_request_headers(
     return {}
 
 
-def contains_anchor(response: Response, anchor: str) -> bool:
+def contains_anchor(response: Response, anchor: str, *, lenient: bool = False) -> bool:
```
Comment (jayaddison) [inline sphinx/builders/linkcheck.py]: I now think that the way to progress this pull request is to:

  * Revert the introduction of any notion of lenient parsing, allowing the removal of the related config setting.
  * Provide a replacement for this helper function named `check_response` or similar.  Initially the replacement function should accept a single, optional `anchor` element -- but it's worth bearing in mind the possibility to check for multiple anchors in a single response to resolve #4303.

Comment (jayaddison) [reply sphinx/builders/linkcheck.py]: (note: I'm not planning to attempt this for 7.4.x - it doesn't seem high priority)

File: CHANGES.rst [hunk 2]
```diff
@@ -145,6 +145,8 @@ Bugs fixed
   Patch by James Addison and Will Lachance.
 * #9634: Do not add a fallback language by stripping the country code.
   Patch by Alvin Wong.
+* #11041: linkcheck: ignore URLs that respond with non-unicode content.
```
Comment (AA-Turner) [inline CHANGES.rst]:
```suggestion
* #11041: linkcheck: Ignore URLs that respond with non-Unicode content.
```

**Comment by jayaddison** [comment]
Thanks @AA-Turner!

**Pull request merged**